src/pykeen/datasets/ea/combination.py

```diff
@@ -44,69 +182,151 @@ def __call__(
         """
+        # concatenate triples
+        mapped_triples, offsets = cat_triples(left, right)
+        # filter alignment and translate to IDs
+        alignment = filter_map_alignment(alignment=alignment, left=left, right=right, entity_offsets=offsets[:, 0])
+        # process
+        mapped_triples, alignment, translation_kwargs = self._process(mapped_triples, alignment, offsets)
+        if isinstance(left, TriplesFactory) and isinstance(right, TriplesFactory):
+            # merge mappings
+            entity_to_id, relation_to_id = _merge_both_mappings(
+                # TODO: offsets can also be dense mapping
+                # translation: Union[torch.LongTensor, Tuple[torch.LongTensor, torch.LongTensor]]
+                left=left,
+                right=right,
+                **translation_kwargs,
+            )
+            triples_factory = TriplesFactory(
+                mapped_triples=mapped_triples,
+                entity_to_id=entity_to_id,
+                relation_to_id=relation_to_id,
+                **kwargs,
+            )
+        else:
+            triples_factory = CoreTriplesFactory(mapped_triples=mapped_triples, **kwargs)
+
+        return triples_factory, alignment
+
+    @abstractmethod
+    def _process(
+        self,
+        mapped_triples: MappedTriples,
+        alignment: torch.LongTensor,
+        offsets: torch.LongTensor,
+    ) -> Tuple[MappedTriples, torch.LongTensor, Mapping[str, Any]]:
         raise NotImplementedError
 
 
+class DisjointGraphPairCombinator(GraphPairCombinator):
+    """This combinator keeps both graphs as disconnected components."""
+
+    # docstr-coverage: inherited
+    def _process(
+        self,
+        mapped_triples: MappedTriples,
+        alignment: torch.LongTensor,
+        offsets: torch.LongTensor,
+    ) -> Tuple[MappedTriples, torch.LongTensor, Mapping[str, Any]]:  # noqa: D102
+        return mapped_triples, alignment, dict(offsets=offsets)
+
+
+class SwapGraphPairCombinator(GraphPairCombinator):
+    """Add extra triples by swapping aligned entities."""
+
+    # docstr-coverage: inherited
+    def _process(
+        self,
+        mapped_triples: MappedTriples,
+        alignment: torch.LongTensor,
+        offsets: torch.LongTensor,
+    ) -> Tuple[MappedTriples, torch.LongTensor, Mapping[str, Any]]:  # noqa: D102
+        # add swap triples
+        # e1 ~ e2 => (e1, r, t) ~> (e2, r, t), or (h, r, e1) ~> (h, r, e2)
+        dense_map = torch.full(size=(offsets[-1, 0],), fill_value=-1)
+        left_id, right_id = alignment
+        dense_map[left_id] = right_id
+        dense_map[right_id] = left_id
+        mapped_triples = torch.cat(
+            [
+                mapped_triples,
+                # swap head
+                _swap_index(mapped_triples=mapped_triples, dense_map=dense_map, index=COLUMN_HEAD),
+                # swap tail
+                _swap_index(mapped_triples=mapped_triples, dense_map=dense_map, index=COLUMN_TAIL),
```

Comment:
@dobraczka I am not sure whether this is the same swapping strategy you implemented.

Reply:
I'm checking it out right now, but I came across some errors, which I'll comment seperately.

I'm using this dummy example to step through the code:
```python
import pandas as pd
left_triples = pd.DataFrame(
    [
        ["la", "0", "lb"],
        ["lb", "0", "lc"],
        ["la", "1", "ld"],
        ["le", "1", "lg"],
        ["lh", "1", "lg"],
    ],
    columns=["head","relation","tail"],
)
left_tf = TriplesFactory.from_labeled_triples(left_triples.values)

right_triples = pd.DataFrame(
    [
        ["ra", "2", "rb"],
        ["ra", "2", "rc"],
        ["rc", "3", "rd"],
        ["re", "3", "rg"],
        ["rh", "3", "rg"],
    ],
    columns=["head","relation","tail"],
)
right_tf = TriplesFactory.from_labeled_triples(right_triples.values)

test_links = pd.DataFrame(
    [
        [left_tf.entity_labeling.label_to_id["ld"],right_tf.entity_labeling.label_to_id["rd"]],
        [left_tf.entity_labeling.label_to_id["le"],right_tf.entity_labeling.label_to_id["re"]],
        [left_tf.entity_labeling.label_to_id["lg"],right_tf.entity_labeling.label_to_id["rg"]],
        [left_tf.entity_labeling.label_to_id["lh"],right_tf.entity_labeling.label_to_id["rh"]],
    ],
    columns=["left","right"],
)
SwapGraphPairCombinator()(left=left_tf,right=right_tf,alignment=test_links)
```

Reply:
added in fe9e4ec...2cb330f